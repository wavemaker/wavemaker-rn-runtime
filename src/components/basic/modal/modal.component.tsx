import React from 'react';
import { View, Modal as ReactModal } from 'react-native';
import { BaseComponent, BaseComponentState } from '@wavemaker/app-rn-runtime/core/base.component';

import WmModalProps from './modal.props';
import { DEFAULT_CLASS, WmModalStyles } from './modal.styles';

export class WmModalState extends BaseComponentState<WmModalProps> {

}

export default class WmModal extends BaseComponent<WmModalProps, WmModalState, WmModalStyles> {

  constructor(props: WmModalProps) {
    super(props, DEFAULT_CLASS, new WmModalProps());
  }

  renderWidget(props: WmModalProps) {
    return React.createElement(View, {
      style: this.styles.root
    }, React.createElement(ReactModal, {
      animationType: props.animationType as any,
      transparent: true,
      //@ts-ignore
      style: this.styles.content
    }, props.children));
  }
}
