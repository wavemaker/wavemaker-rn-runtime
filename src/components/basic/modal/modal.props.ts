import { BaseProps } from '@wavemaker/app-rn-runtime/core/base.component';

export default class WmModalProps extends BaseProps {
    children = null as any;
    animationType? = 'slide';
}