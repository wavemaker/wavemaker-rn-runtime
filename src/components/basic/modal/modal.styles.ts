import BASE_THEME, { AllStyle }  from '@wavemaker/app-rn-runtime/styles/theme';
import { BaseStyles } from '@wavemaker/app-rn-runtime/core/base.component';


export type WmModalStyles = BaseStyles & {
    content: AllStyle,
};

export const DEFAULT_CLASS = 'app-modal';
export const DEFAULT_STYLES: WmModalStyles = {
    root: {
        backgroundColor: 'rgba(0, 0, 0, 0)'
    },
    text: {},
    content: {
        borderColor: 'rgba(0, 0, 0, 0)',
        borderWidth: 0
    }
};

BASE_THEME.addStyle('centered-modal', '', {
    root: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        alignSelf: 'center',
    },
    text: {},
    content: {
        alignSelf: 'center'
    }
} as WmModalStyles);


BASE_THEME.addStyle(DEFAULT_CLASS, '', DEFAULT_STYLES);