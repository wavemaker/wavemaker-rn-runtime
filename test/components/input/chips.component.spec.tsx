import React, { ReactNode, createRef } from 'react';
import { Text, TouchableOpacity } from 'react-native';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react-native';
import WmChips from '@wavemaker/app-rn-runtime/components/input/chips/chips.component';
import WmChipsProps from '../../../src/components/input/chips/chips.props';

import WmIcon from '@wavemaker/app-rn-runtime/components/basic/icon/icon.component';
import WmSearch from '@wavemaker/app-rn-runtime/components/basic/search/search.component';
import WmPicture from '@wavemaker/app-rn-runtime/components/basic/picture/picture.component';

import { ModalProvider } from '@wavemaker/app-rn-runtime/core/modal.service';
import { AssetProvider } from '@wavemaker/app-rn-runtime/core/asset.provider';
import AppModalService from '@wavemaker/app-rn-runtime/runtime/services/app-modal.service';

const timer = (time = 100) =>
  new Promise((resolve: any, reject) => {
    setTimeout(() => resolve(), time);
  });

const appConfig = {
  app: {
    toastsOpened: 1,
  },
  refresh: () => {},
};
const loadAsset = (path) => path;

AppModalService.modalsOpened = [];

jest.mock('@wavemaker/app-rn-runtime/core/injector', () => {
  const actualInjector = jest.requireActual(
    '@wavemaker/app-rn-runtime/core/injector'
  );
  return {
    ...actualInjector,
    get: jest.fn().mockImplementation(() => {
      return appConfig;
    }),
    FOCUSED_ELEMENT: {
      get: jest.fn().mockImplementation(() => ({
        blur: jest.fn(),
      })),
    },
  };
});

describe('WmChips', () => {
  let defaultProps: WmChipsProps;
  let datasetProps;

  beforeEach(() => {
    defaultProps = new WmChipsProps();
    defaultProps.dataset = [
      {
        name: 'name0',
        dataValue: 'dataValue0',
      },
      {
        name: 'name1',
        dataValue: 'dataValue1',
      },
      {
        name: 'name2',
        dataValue: 'dataValue2',
      },
    ];
    defaultProps.searchable = true;
    defaultProps.placeholder = 'Search chips...';
    defaultProps.searchkey = 'name';

    datasetProps = {
      datafield: 'name',
      displayfield: 'name',
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  function renderComponentWithWrappers(props = {}) {
    const ref = createRef();
    return render(
      <ModalProvider value={AppModalService}>
        <AssetProvider value={loadAsset}>
          <WmChips {...defaultProps} {...datasetProps} {...props} />
        </AssetProvider>
      </ModalProvider>
    );
  }

  const invokeEventCallbackMock = jest.spyOn(
    WmChips.prototype,
    'invokeEventCallback'
  );

  // Check Rendering with Default Props
  it('renders correctly with default props and type as search', () => {
    render(<WmChips {...defaultProps} {...datasetProps} />);
    expect(screen.getByPlaceholderText('Search chips...')).toBeTruthy();
    let searchInput = screen.UNSAFE_getByType(WmSearch);
    expect(searchInput.props.type).toBe('search');
  });

  it('renders correctly with default props and type as autocomplete', async () => {
    render(<WmChips {...defaultProps} {...datasetProps} minchars={0} />);
    expect(screen.getByPlaceholderText('Search chips...')).toBeTruthy();
    let searchInput = screen.UNSAFE_getByType(WmSearch);
    expect(searchInput.props.type).toBe('autocomplete');
  });

  // Dataset Handling
  it('handles dataset properly and renders chips based on it', () => {
    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        searchable={false}
        datavalue="name1"
      />
    );

    const selectedChip = screen.getByText('name1');
    expect(screen.getByText('name0')).toBeTruthy();
    expect(screen.getByText('name2')).toBeTruthy();
    expect(selectedChip).toBeTruthy();
    expect(selectedChip.parent?.parent?.props.accessibilityState.selected).toBe(
      true
    );

    // check selected color as well
  });

  // Prop Update Handling
  it('responds to prop updates correctly', async () => {
    const { rerender } = render(
      <WmChips {...defaultProps} {...datasetProps} />
    );
    const newDataset = [{ name: 'Chip 3', dataValue: 'chip3' }];
    rerender(
      <WmChips
        {...defaultProps}
        dataset={newDataset}
        {...datasetProps}
        searchable={false}
      />
    );
    await waitFor(() => {
      expect(screen.getByText('Chip 3')).toBeTruthy();
    });
  });

  // Item Addition
  it('adds chips correctly', async () => {
    // AppModalService.modalsOpened = [];

    const tree = renderComponentWithWrappers();
    const searchInput = tree.getByPlaceholderText('Search chips...');
    const search = tree.UNSAFE_getByType(WmSearch);
    search.instance.view = {
      measure: (callback: Function) => {
        callback(0, 0, 100, 50, 10, 20);
      },
    } as any;
    fireEvent(searchInput, 'onFocus');
    fireEvent.changeText(searchInput, 'name2');

    await timer(500);

    const renderOptions = AppModalService.modalOptions; //[0];
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);

    const selectItem = subTree.getByText('name2');
    expect(selectItem).toBeTruthy();

    fireEvent.press(selectItem);
    await waitFor(() => {
      expect(tree.getByText('name2')).toBeTruthy();
    });
    // should not add same chip if selected again.
    fireEvent.press(selectItem);
    const resetSearchModelMock = jest.spyOn(
      WmChips.prototype,
      'resetSearchModel'
    );
    await waitFor(() => {
      expect(resetSearchModelMock).toHaveBeenCalled();
    });
  });

  // Item Removal
  it('removes chips correctly', async () => {
    render(<WmChips {...defaultProps} {...datasetProps} datavalue="name2" />);
    fireEvent.press(screen.getByText('clear'));
    const removeItemMock = jest.spyOn(WmChips.prototype, 'removeItem');
    await waitFor(() => {
      expect(removeItemMock).toHaveBeenCalled();
    });
    expect(screen.queryByText('name2')).toBeNull();
  });

  // Max Size Handling
  it('does not add more chips when max size is reached', async () => {
    // AppModalService.modalsOpened = [];
    const tree = renderComponentWithWrappers({
      maxsize: 1,
      datavalue: 'name2',
    });
    const searchInput = tree.getByPlaceholderText('Search chips...');
    const search = tree.UNSAFE_getByType(WmSearch);
    search.instance.view = {
      measure: (callback: Function) => {
        callback(0, 0, 100, 50, 10, 20);
      },
    } as any;
    fireEvent(searchInput, 'onFocus');
    fireEvent.changeText(searchInput, 'name');

    await timer(500);

    const renderOptions = AppModalService.modalOptions; //[0];
    const Content = () => {
      return <>{renderOptions.content}</>;
    };
    const subTree = render(<Content />);
    fireEvent.press(subTree.getByText('name1'));
    await timer(300);
    await waitFor(() => {
      expect(tree.getByText('name2')).toBeTruthy();
      // expect(tree.getByText('name1')).toBeTruthy();
      expect(tree.queryByText('name1')).toBeNull();
    });
  });

  // Accessibility Props
  it('applies accessibility props correctly', async () => {
    const ref = createRef();
    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        searchable={false}
        ref={ref}
        accessibilitylabel="Chips Component"
        hint="chips"
      />
    );

    await waitFor(() => {
      // expect(screen).toMatchSnapshot();
      expect(screen.getByLabelText('Chips Component')).toBeTruthy();
      expect(screen.getByAccessibilityHint('chips')).toBeTruthy();
    });
    // console.log('state.props ===', ref.current.state);
  });

  it('should be able to select chip when searchable is false', () => {
    const ref = createRef();
    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        searchable={false}
        ref={ref}
      />
    );
    const selectChipMock = jest.spyOn(WmChips.prototype, 'selectChip');
    const setDatavalueMock = jest.spyOn(WmChips.prototype, 'setDatavalue');
    const chipItem = screen.getByText('name1');
    fireEvent.press(chipItem);
    expect(selectChipMock).toHaveBeenCalled();
    expect(setDatavalueMock).toHaveBeenCalled();
    expect(invokeEventCallbackMock).toHaveBeenCalled();
    expect(ref.current.state.dataItems[1].selected).toBe(true);
  });

  it('should handle readonly properly', async () => {
    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        datavalue="name1"
        readonly={true}
      />
    );
    fireEvent.press(screen.getByText('clear'));
    await timer(300);
    await waitFor(() => {
      expect(screen.getByText('name1')).toBeTruthy();
    });

    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        datavalue="name1"
        readonly={true}
        searchable={false}
      />
    );
    const defaultChip = screen.getByText('name1');
    fireEvent.press(defaultChip);
    await timer(300);
    expect(defaultChip.parent?.parent?.props.accessibilityState.selected).toBe(
      true
    );
  });

  it('should handle disabled properly', async () => {
    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        datavalue="name1"
        disabled={true}
      />
    );
    expect(screen.queryByText('clear')).toBeNull();
    fireEvent.press(screen.getByText('name1'));

    await waitFor(() => {
      expect(screen.getByText('name1')).toBeTruthy();
    });
  });

  it('should not allow to select chips more than maxsize', async () => {
    render(
      <WmChips
        {...defaultProps}
        {...datasetProps}
        searchable={false}
        datavalue="name1"
        maxsize={1}
      />
    );
    const defaultChip = screen.getByText('name1');
    const selectedChip = screen.getByText('name2');
    fireEvent.press(selectedChip);
    expect(selectedChip.parent?.parent?.props.accessibilityState.selected).toBe(
      false
    );
    expect(defaultChip.parent?.parent?.props.accessibilityState.selected).toBe(
      true
    );

    fireEvent.press(defaultChip);
    await timer();
    fireEvent.press(selectedChip);

    await waitFor(() => {
      expect(
        selectedChip.parent?.parent?.props.accessibilityState.selected
      ).toBe(true);
    });
  });
});
