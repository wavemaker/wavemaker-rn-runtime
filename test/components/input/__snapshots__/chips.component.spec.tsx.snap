// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WmChips renders correctly with default props and type as search 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "flexWrap": "wrap",
    }
  }
>
  <View
    accessibilityLabel="Choose tags"
    accessible={true}
    style={
      {
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
  >
    <View
      style={
        [
          {
            "flexDirection": "column",
            "width": "100%",
          },
          {
            "flexDirection": "column-reverse",
          },
        ]
      }
    >
      <View
        onLayout={[Function]}
        style={
          [
            undefined,
          ]
        }
      >
        <View
          onLayout={[Function]}
          style={
            {
              "borderRadius": 28,
              "flexDirection": "row",
              "marginBottom": 8,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "center",
              }
            }
          >
            <TextInput
              accessibilityRole="search"
              accessible={true}
              autoFocus={false}
              defaultValue=""
              editable={true}
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onLayout={[Function]}
              placeholder="Search chips..."
              placeholderTextColor="#bbbbbb"
              style={
                [
                  {
                    "backgroundColor": "#dddddd",
                    "borderBottomLeftRadius": 28,
                    "borderBottomRightRadius": 4,
                    "borderColor": "#dddddd",
                    "borderRightWidth": 1,
                    "borderTopLeftRadius": 28,
                    "borderTopRightRadius": 4,
                    "borderWidth": 1,
                    "fontFamily": "Roboto",
                    "minHeight": 56,
                    "paddingBottom": 8,
                    "paddingLeft": 24,
                    "paddingRight": 14,
                    "paddingTop": 8,
                    "userSelect": "text",
                    "width": "100%",
                  },
                  {},
                  null,
                ]
              }
              testID="null_search_i"
            />
          </View>
        </View>
      </View>
      <View
        accessibilityLabel="Choose tags"
        accessible={true}
        style={
          {
            "flexDirection": "row",
            "flexWrap": "wrap",
          }
        }
      />
    </View>
  </View>
</View>
`;
